### ISSUE TITLE
ChatGPT Solution for our FanAOI project

### ISSUE THREAD

**Issue by Austin-Chang-zz** [orange811024/FanAOI#7]
### Step 1: Define the Custom Filter Matrix
To detect horizontal lines, we can use filters that emphasize horizontal edges. The Sobel filter for horizontal edges is a good example:

#### Sobel Filter for Horizontal Edges:
```python
horizontal_edge_filter = np.array([
    [-1, -2, -1],
    [ 0,  0,  0],
    [ 1,  2,  1]
], dtype='float32')
```

Alternatively, if you want to emphasize thicker horizontal lines like the ones in your label, you can tweak this filter:

#### Custom Filter for Thicker Horizontal Lines:
```python
horizontal_line_filter = np.array([
    [-1, -1, -1],
    [ 2,  2,  2],
    [-1, -1, -1]
], dtype='float32')
```

These filters will detect horizontal lines effectively, and we will use them in the CNN model.

### Step 2: CNN for Label Inspection

We'll define a CNN that takes these filters and is trained to classify images as "GO" (acceptable) or "NG" (not acceptable).

### Step 3: CNN Model Code

Here's a Python script to define and train a CNN model for detecting tilted labels:

```python
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
from tensorflow.keras.preprocessing.image import ImageDataGenerator
import matplotlib.pyplot as plt

# Define a custom filter for detecting horizontal lines
horizontal_line_filter = np.array([
    [-1, -1, -1],
    [ 2,  2,  2],
    [-1, -1, -1]
], dtype='float32').reshape((3, 3, 1, 1))  # Reshape for single-channel input and output

# Build the CNN model
model = Sequential()

# Add convolutional layer with custom filter
model.add(Conv2D(filters=1, kernel_size=(3, 3), activation='relu', input_shape=(64, 64, 1)))
model.layers[0].set_weights([horizontal_line_filter, np.array([0.0])])  # Initialize with custom filter

# Add pooling layer
model.add(MaxPooling2D(pool_size=(2, 2)))

# Add another convolutional layer
model.add(Conv2D(filters=32, kernel_size=(3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))

# Flatten the results and add a dense layer for classification
model.add(Flatten())
model.add(Dense(units=128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(units=2, activation='softmax'))  # 2 classes: GO and NG

# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Summary of the model
model.summary()

# Assuming you have images preprocessed and split into training and testing sets
# Example:
# training_set, testing_set = <your preprocessing and splitting code>

# Train the model
# model.fit(training_set, epochs=10, validation_data=testing_set)

# Save the model
model.save('cooling_fan_label_inspection.h5')
```

### Step 4: GO/NG Classification Code

After training the model, use it to predict and classify new images:

```python
from tensorflow.keras.models import load_model
import cv2

# Load the trained model
model = load_model('cooling_fan_label_inspection.h5')

def classify_image(image_path):
    # Load and preprocess the image
    img = cv2.imread(image_path, cv2.IMREAD_GRAYSCALE)
    img = cv2.resize(img, (64, 64))
    img = img.reshape(1, 64, 64, 1) / 255.0

    # Predict using the trained model
    prediction = model.predict(img)
    class_idx = np.argmax(prediction)

    return "GO" if class_idx == 0 else "NG"

# Example usage
result = classify_image('path_to_image.png')
print("Classification Result:", result)
```

### Summary
1. **Step 1**: Define filters for horizontal lines.
2. **Step 2**: Train a CNN using these filters and label data.
3. **Step 3**: Use the trained CNN to classify new images.
4. **Step 4**: Classify each image as "GO" or "NG" based on model output.

Make sure to preprocess your images (grayscale conversion, resizing, normalization) and split them into training and validation sets before training the CNN. Let me know if you need further assistance with preprocessing or any other part!
